geode-core/src/main/java/org/apache/geode/internal/cache/InternalCache.java

```diff
@@ -148,6 +148,8 @@
 
   QueryService getLocalQueryService();
 
+  <K, V> Region<K, V> getQueryStore(String storeName);
```

Comment:
I disagree.  InternalCacheForClientAccess implements InternalCache and should be indistinguishable from other InternalCache implementations.  I liked your first idea because it removed knowledge of server vs client cache access from the REST code.
Having getQueryStore take an argument is more flexible and leaves it open to allowing there to be multiple query stores.  Since that's the way the REST code was written I think it should stay like that.

Reply:
I would consider changing the method name from getQueryStore to getInternalRegion though.  I think that would make the JMX code that I'm now working on cleaner because I could install an InternalCacheForClientAccess into it and then use getInternalRegion in specific places where it needs to get its "monitoring regions"

Reply:
but then wouldn't the  client just pass in any internal region name to that method and get access to any internal region? I believe in the REST code, any caller to that method is passing in the same string... (but I may have missed something).

In my thought, CacheForClientAccess is a child of InternalCache which has restricted access to the regions, but aside from that, it needs to access some "internal regions" for legacy reasons. InternalCache already has that capability in the "getRegion" method, so it doesn't need anymore method to get a special internal region. But for CacheClientAccess access, some special internal region will be allowed access by adding some additional method. My 2 cents.

Reply:
This isn't code that's accessible from a client.  This is server-side code and it's okay for that code to intentionally get an internal region.  What we don't want is someone using the REST API to be able to send a request that modifies an internal region or returns privileged information from an internal region.

Reply:
Ah, I think if I have the REST code assume it has an InternalCacheForClientAccess instead of an InternalCache I can remove the getQueryStore method from InternalCache and GemFireCacheImpl.  Yeah, that will work.  Thanks Jinmei

Reply:
Yes, that's what I meant as well. The rest code should be working with InternalCacheForClientAccess alone because that is what it needs. I like the renaming that method to getInterRegion for a more general purpose.